service/history/workflow/mutable_state_impl.go

```diff
@@ -5012,6 +5012,7 @@ func (ms *MutableStateImpl) GetUpdateCondition() (int64, int64) {
 func (ms *MutableStateImpl) SetSpeculativeWorkflowTaskTimeoutTask(
 	task *tasks.WorkflowTaskTimeoutTask,
 ) error {
+	task.InMemory = true
```

Comment:
Can you please move it one level up in call stack? `SetSpeculativeWorkflowTaskTimeoutTask` is called two times and this is where task is constructed. All fields assignment should be as close as possible to it (i.e. task is ~immutable).